src/hypothesis/extra/numpy.py

```diff
@@ -56,12 +72,19 @@ class ArrayStrategy(SearchStrategy):
     def __init__(self, element_strategy, shape, dtype):
         self.shape = tuple(shape)
         assert shape
-        self.array_size = reduce(operator.mul, shape)
+        self.array_size = np.prod(shape)
+        if self.array_size * dtype.itemsize > settings().buffer_size:
+            raise ValueError((
```

Comment:
I'll switch out all the asserts, but is a value drawn from a strategy conceptually an invalid argument?  (I lean towards yes, but worth checking)